File: frontend/src/features/imports/api/imports.ts
/**
 * Import API contract.
 *
 * Schemas mirror `backend/app/schemas/imports.py`. A submission may carry pasted text,
 * one file, or several files together — the backend treats a single-game PGN as the N=1
 * case of the same batch path, so there is one `createImport` call, not a separate one
 * for "single" vs "batch".
 */
import { z } from 'zod';

import { apiClient } from '@/shared/lib/api-client';

const rejectedGameSchema = z.object({
  source: z.string(),
  index: z.number(),
  reason: z.string(),
  detail: z.string(),
});

const jobProgressSchema = z.object({
  total: z.number().default(0),
  imported: z.number().default(0),
  duplicates: z.number().default(0),
  rejected: z.array(rejectedGameSchema).default([]),
});

export const jobSummarySchema = z.object({
  id: z.string(),
  kind: z.string(),
  status: z.enum(['pending', 'processing', 'done', 'failed']),
  progress: jobProgressSchema,
  error: z.record(z.string(), z.unknown()).nullable(),
  created_at: z.string(),
  completed_at: z.string().nullable(),
});
export type JobSummary = z.infer<typeof jobSummarySchema>;
export type RejectedGame = z.infer<typeof rejectedGameSchema>;

const jobListSchema = z.array(jobSummarySchema);

/** Submit pasted PGN text, uploaded files, or both together as one import job. */
export function createImport(input: { pgnText?: string; files?: File[] }): Promise<JobSummary> {
  const form = new FormData();
  if (input.pgnText && input.pgnText.trim()) {
    form.append('pgn_text', input.pgnText);
  }
  for (const file of input.files ?? []) {
    form.append('files', file);
  }
  return apiClient.post('/api/v1/imports', jobSummarySchema, form);
}

export function fetchImportJob(jobId: string, signal?: AbortSignal): Promise<JobSummary> {
  return apiClient.get(`/api/v1/imports/${jobId}`, jobSummarySchema, signal);
}

export function fetchImportJobs(signal?: AbortSignal): Promise<JobSummary[]> {
  return apiClient.get('/api/v1/imports', jobListSchema, signal);
}

/**
 * Sync recent games from Lichess/Chess.com (Phase 14, D-030/D-031).
 *
 * Returns immediately with a `pending` job — the platform fetch and ingestion run in
 * the background server-side, so callers should hand the returned job id to
 * `ImportJobStatus`/`useImportJob`, the same polling path a manual upload's job uses.
 *
 * `username` omitted syncs the caller's own linked account. Supplying one imports a
 * player being studied; the backend routes those games to the study profile on its own
 * (see `PlatformSyncRequest`), so there is no target-profile argument here.
 */
export function syncFromPlatform(
  provider: 'lichess' | 'chesscom',
  window?: number,
  username?: string,
): Promise<JobSummary> {
  return apiClient.post(`/api/v1/imports/${provider}/sync`, jobSummarySchema, {
    window,
    username,
  });
}

File: backend/app/schemas/imports.py
"""Import request and response schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class RejectedGameSummary(BaseModel):
    """One game that did not make it in, and why."""

    source: str
    index: int
    reason: str
    detail: str


class JobProgress(BaseModel):
    total: int = 0
    imported: int = 0
    duplicates: int = 0
    rejected: list[RejectedGameSummary] = []


class JobSummary(BaseModel):
    """An import job's visible status. Returned by create, get, and list."""

    id: uuid.UUID
    kind: str
    status: str
    progress: JobProgress
    error: dict[str, Any] | None
    created_at: datetime
    completed_at: datetime | None


class PlatformSyncRequest(BaseModel):
    """Sync request body (Phase 14). `window` is optional — the route falls back to
    `AnalyticsSettings.analytics_default_window`, the same "how many recent games"
    default the profile-analytics window picker already uses (D-030/D-031: no new
    window-size concept is introduced for imports)."""

    window: int | None = None
    # Phase 16b follow-up: whose games to fetch. Omitted means the caller's own linked
    # account for this provider (Phase 14's only behaviour). Supplied means an arbitrary
    # player being studied — the games land in the caller's study profile on their own,
    # because `ImportService._target_profile_id` already routes per game on "do these
    # headers match a linked username of mine" (D-021, ADR-0016). No target-profile
    # field is needed here, and adding one would duplicate that routing decision.
    username: str | None = None


__all__ = ["JobProgress", "JobSummary", "PlatformSyncRequest", "RejectedGameSummary"]
